apps/passport-client/src/dispatch.ts

```diff
@@ -942,3 +948,36 @@ async function promptToAgreePrivacyNotice(state: AppState, update: ZuUpdate) {
     });
   }
 }
+
+/**
+ * Formats the user's data as JSON, then creates a data URL link and
+ * simulates a click on that link in order to trigger a download.
+ */
+async function exportData(state: AppState, _update: ZuUpdate) {
+  const { self: user, pcds, subscriptions } = state;
+
+  // Since we already use this data for remote sync, we know that it's
+  // sufficient for loading an account on to a new device.
+  const { serializedStorage, storageHash } = await serializeStorage(
+    user,
+    pcds,
+    subscriptions
+  );
+
+  // Data in a data URL must be Base64-encoded
+  const data = Buffer.from(JSON.stringify(serializedStorage)).toString(
+    "base64"
+  );
+
+  const dataURL = `data://text/json;base64,${data}`;
+
+  const link = document.createElement("a");
+  link.href = dataURL;
+  // Give the browser a meaningful filename to use for the file
+  link.download = `zupass-${storageHash}.json`;
+  // Trigger download
+  link.click();
+  link.remove();
```

Comment:
thought: I am surprised that this works. are you sure this is best practice?

Reply:
What we're doing is a little bit unusual - we're trying to save some local data, without doing a server-side fetch/download. There is support for this using the [File System API](https://developer.mozilla.org/en-US/docs/Web/API/File_System_API) but this is not fully supported in all browsers, and requires permissions that we don't actually need (there's a discussion of this [here](https://stackoverflow.com/questions/66294683/how-to-implement-save-and-save-as-without-download-using-javascript - diagrams.net was the example I had in mind for arbitrary file-saving).

This means that we want the browser to "download" a file, and the data URL approach seems to be commonplace. One other option might be to have a "magic" URL that is intercepted by the service worker, and which returns a result based on local storage without actually calling out to the server (which doesn't have the necessary information anyway).

On reflection, I think rather than doing a `dispatch` and generating the link tag manually, I might just be able to make the button the user clicks on a link tag which links directly to the data URL, which would simplify the code a bit.

Reply:
Since service workers can't access local storage, triggering the download by pointing to a URL managed by the service worker wouldn't work either.

Reply:
OK, just checking.